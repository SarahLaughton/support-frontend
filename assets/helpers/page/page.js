// @flow

// ----- Imports ----- //

import { createStore, combineReducers } from 'redux';
import 'ophan';

import * as ga from 'helpers/tracking/ga';
import * as abTest from 'helpers/abtest';
import * as logger from 'helpers/logger';
import { getCampaign } from 'helpers/tracking/guTracking';
import { getQueryParameter } from 'helpers/url';
import { detect } from 'helpers/internationalisation/country';

import type { Campaign } from 'helpers/tracking/guTracking';
import type { IsoCountry } from 'helpers/internationalisation/country';
import type { Participations } from 'helpers/abtest';

import type { Action } from './pageActions';


// ----- Types ----- //

type CommonState = {
  intCmp: ?string,
  campaign: ?Campaign,
  refpvid: ?string,
  country: IsoCountry,
  abParticipations: Participations,
};


// ----- Functions ----- //

// Sets up GA and logging.
function analyticsInitialisation(participations: Participations): void {

  // Google analytics.
  ga.init();
  ga.setDimension('experience', abTest.getVariantsAsString(participations));
  ga.trackPageview();

  // Logging.
  logger.init();

}

// Creates the initial state for the common reducer.
function buildInitialState(abParticipations: Participations) {

  const intCmp = getQueryParameter('INTCMP');

  const initialState: CommonState = {
    intCmp,
    campaign: intCmp ? getCampaign(intCmp) : null,
    refpvid: getQueryParameter('REFPVID'),
    country: detect(),
    abParticipations,
  };

  return initialState;

}

// Sets up the common reducer with its initial state.
function createCommonReducer(
  initialState: CommonState): (CommonState, Action) => CommonState {

  function commonReducer(
    state: CommonState = initialState,
    action: Action): CommonState {

    switch (action.type) {

      case 'SET_INTCMP':
        return Object.assign({}, state, {
          intCmp: action.intCmp,
          campaign: getCampaign(action.intCmp),
        });

      case 'SET_COUNTRY':
        return Object.assign({}, state, { country: action.country });

      default:
        return state;

    }

  }

  return commonReducer;

}

// Initialises the page.
function init(pageReducer: Object, preloadedState: ?Object, middleware: ?Function) {

  const participations: Participations = abTest.init();
  analyticsInitialisation(participations);
  const initialState: CommonState = buildInitialState(participations);
  const commonReducer = createCommonReducer(initialState);

  return createStore(
    combineReducers({ page: pageReducer, common: commonReducer }),
    preloadedState,
    middleware,
  );

}


// ----- Exports ----- //

export {
  createCommonReducer,
  init,
};
