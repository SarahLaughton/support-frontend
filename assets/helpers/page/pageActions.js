// @flow

// ----- Imports ----- //

import type { IsoCountry } from 'helpers/internationalisation/country';


// ----- Types ----- //

export type Action =
  | { type: 'SET_INTCMP', intCmp: string }
  | { type: 'SET_COUNTRY', country: IsoCountry }
  ;


// ----- Action Creators ----- //

export function setIntCmp(intCmp: string): Action {
  return { type: 'SET_INTCMP', intCmp };
}

export function setCountry(country: IsoCountry): Action {
  return { type: 'SET_COUNTRY', country };
}
