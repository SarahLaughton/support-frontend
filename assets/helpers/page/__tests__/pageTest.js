// @flow

// ----- Imports ----- //

import type { Country } from 'helpers/internationalisation/country';

import { createCommonReducer } from '../page';


// ----- Tests ----- //

jest.mock('ophan', () => {});

describe('reducer tests', () => {

  let reducer = () => {};

  beforeEach(() => {

    const initialState = {
      intCmp: null,
      campaign: 'dummy_campaign',
      refpvid: null,
      country: 'GB',
      abParticipations: {},
    };

    reducer = createCommonReducer(initialState);

  });

  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot();
  });

  it('should handle SET_INTCMP to "dummy_intcmp"', () => {

    const intCmp = 'dummy_intcmp';
    const action = {
      type: 'SET_INTCMP',
      intCmp,
    };

    const newState = reducer(undefined, action);

    expect(newState.intCmp).toEqual(intCmp);
    expect(newState.campaign).toBeNull();
    expect(newState.refpvid).toMatchSnapshot();
    expect(newState.country).toMatchSnapshot();
    expect(newState.abParticipations).toMatchSnapshot();
  });

  it('should handle SET_COUNTRY to US', () => {

    const country: Country = 'US';
    const action = {
      type: 'SET_COUNTRY',
      country,
    };

    const newState = reducer(undefined, action);

    expect(newState.country).toEqual(country);
    expect(newState.intCmp).toMatchSnapshot();
    expect(newState.campaign).toMatchSnapshot();
    expect(newState.refpvid).toMatchSnapshot();
    expect(newState.abParticipations).toMatchSnapshot();
  });

});
