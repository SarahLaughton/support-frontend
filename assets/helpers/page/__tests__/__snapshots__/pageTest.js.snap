// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer tests should handle SET_COUNTRY to US 1`] = `null`;

exports[`reducer tests should handle SET_COUNTRY to US 2`] = `"dummy_campaign"`;

exports[`reducer tests should handle SET_COUNTRY to US 3`] = `null`;

exports[`reducer tests should handle SET_COUNTRY to US 4`] = `Object {}`;

exports[`reducer tests should handle SET_INTCMP to "dummy_intcmp" 1`] = `null`;

exports[`reducer tests should handle SET_INTCMP to "dummy_intcmp" 2`] = `"GB"`;

exports[`reducer tests should handle SET_INTCMP to "dummy_intcmp" 3`] = `Object {}`;

exports[`reducer tests should return the initial state 1`] = `
Object {
  "abParticipations": Object {},
  "campaign": "dummy_campaign",
  "country": "GB",
  "intCmp": null,
  "refpvid": null,
}
`;
